common/setups/returnn_common/nnet_constructor.py

```diff
@@ -0,0 +1,425 @@
+
+Usage Example:
+
+    rc_some_unhashed_code = NonhashedCode(code=some_code)
+    rc_extern_data = ReturnnCommonExternData(extern_data=extern_data)
+    rc_model = ReturnnCommonImport(
+        "i6_experiments.users.some_user.common_modules.some_module.SomeNnModuleModel")
+    rc_construction_code = ReturnnCommonImport(
+        "i6_experiments.users.some_user.common_modules.constructor_module.some_constructor_function")
+
+    rc_network = ReturnnCommonDynamicNetwork(
+        net_func_name=rc_construction_code.get_name(),
+        net_func_map={"net_module": rc_model.get_name(),
+                      "audio_data": "audio_features",
+                      "label_data": "bpe_labels",
+                      "audio_feature_dim": "audio_features_feature",
+                      "audio_time_dim": "audio_features_time",
+                      "label_time_dim": "bpe_labels_time",
+                      "label_dim": "bpe_labels_indices"
+                     },
+        net_kwargs={... some additional constructor function args ...}
+    )
+
+    serializer = ReturnnCommonSerializer(
+        serializer_objects=[rc_recursionlimit,
+                            rc_extern_data,
+                            rc_model,
+                            rc_construction_code,
+                            rc_network],
+        returnn_common_root=returnn_common_root,
+        make_local_package_copy=True, # set to false to directly import from the recipes
+    )
+    returnn_config = ReturnnConfig(
+        config=config,
+        post_config=post_config,
+        python_epilog=[serializer],
+    )
+"""
+from dataclasses import dataclass, asdict
+from typing import Any, List, Union, Optional, Dict
+import os
+import pathlib
+import shutil
+import string
+import textwrap
+
+from sisyphus import tk
+from sisyphus.delayed_ops import DelayedBase
+from sisyphus.hash import sis_hash_helper
+
+from i6_core.util import instanciate_delayed, uopen
+from i6_core.returnn.config import CodeWrapper
+
+
+class SerializerObject(DelayedBase):
+    """
+    Any class that can be passed to ReturnnCommonSerializer
+    """
+
+    def __init__(self):
+        # suppress init warning
+        super().__init__(None)
+
+
+@dataclass(frozen=True)
+class DimInitArgs:
+    """
+    A helper class to store input args for a nn.Dim object
+
+    :param name: name of the dim
+    :param dim: dimension size (feature axis size or label index count for sparse_dim), None for spatial axis
+    """
+
+    name: str
+    dim: Optional[Union[int, tk.Variable]]
+    is_feature: bool = False
+
+
+@dataclass(frozen=True)
+class DataInitArgs:
+    """
+    A helper class to store needed input args for a nn.Data object without `returnn_common` dependency
+
+    :param name: name of the data (equivalent to the extern_data entry)
+    :param available_for_inference: if this data is available during decoding/forward pass etc...
```

Comment:
I dont really like the wording here, but I dont like the RETURNN Data class docstring either" `available_for_inference: e.g. the extern data "classes" is usually not available for inference`. 
The RETURNN docs say `available_for_inference defines if the data stream is also available during the “search” or “forward” task, so basically marking if a data stream is an input or only a target.` 
So maybe something like `defines if the data is also available during “search” or “forward`, maybe even change on RETURNN side too. What do you think @albertz

Reply:
lets postpone this discussion, I will just merge so that we can start working with this code.